File: apps/ios/Zeron/Auth/AuthClient.swift
// Edge auth client — /auth/exchange, /auth/refresh, /auth/orgs
// (edge/src/auth-routes.ts). Two modes, mirroring the engine:
// - WorkOS: paste-code exchange → access/refresh tokens; refresh scoped to an
//   org adds the org_id claim the workspace room requires.
// - Dev (AUTH_MODE=dev edge): the bearer string IS the user id; "user@org"
//   supplies a fake org claim.

import Foundation
import Security

struct AuthUser: Codable, Equatable {
    var id: String
    var email: String?
    var firstName: String?
    var lastName: String?
}

struct AuthOrg: Codable, Identifiable, Equatable {
    var id: String
    var organizationId: String
    var name: String
}

struct AuthTokens: Codable, Equatable {
    var accessToken: String
    var refreshToken: String
}

enum AuthError: LocalizedError {
    case http(Int, String)
    case invalidResponse

    var errorDescription: String? {
        switch self {
        case .http(let code, let body): return "Auth failed (\(code)): \(body)"
        case .invalidResponse: return "Unexpected auth response"
        }
    }
}

struct AuthClient {
    var baseURL: URL

    func exchange(code: String) async throws -> (AuthUser, AuthTokens) {
        struct Response: Codable {
            var user: AuthUser
            var accessToken: String
            var refreshToken: String
        }
        let r: Response = try await post("auth/exchange", body: ["code": code])
        return (r.user, AuthTokens(accessToken: r.accessToken, refreshToken: r.refreshToken))
    }

    func refresh(refreshToken: String, organizationId: String? = nil) async throws -> AuthTokens {
        var body: [String: String] = ["refreshToken": refreshToken]
        if let organizationId { body["organizationId"] = organizationId }
        return try await post("auth/refresh", body: body)
    }

    func orgs(accessToken: String) async throws -> [AuthOrg] {
        struct Response: Codable { var orgs: [AuthOrg] }
        var request = URLRequest(url: baseURL.appending(path: "auth/orgs"))
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.check(data: data, response: response)
        return try JSONDecoder().decode(Response.self, from: data).orgs
    }

    private func post<T: Decodable>(_ path: String, body: [String: String]) async throws -> T {
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(body)
        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.check(data: data, response: response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private static func check(data: Data, response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { throw AuthError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            throw AuthError.http(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
    }
}

// MARK: - Keychain storage

enum Keychain {
    private static let service = "sh.zeron.ios"

    static func save(_ value: String, key: String) {
        _ = saveChecked(value, key: key)
    }

    @discardableResult
    static func saveChecked(_ value: String, key: String, deviceOnly: Bool = false) -> Bool {
        let data = Data(value.utf8)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(query as CFDictionary)
        var add = query
        add[kSecValueData as String] = data
        add[kSecAttrAccessible as String] = deviceOnly
            ? kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly : kSecAttrAccessibleAfterFirstUnlock
        return SecItemAdd(add as CFDictionary, nil) == errSecSuccess
    }

    static func load(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
              let data = result as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func delete(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

File: apps/android/app/src/main/java/sh/zeron/android/auth/AuthClient.kt
package sh.zeron.android.auth

import org.json.JSONObject
import sh.zeron.android.config.AppConfig
import sh.zeron.android.core.AppError
import sh.zeron.android.sync.HttpTransport

data class AuthUser(val id: String, val email: String?)
data class AuthOrg(val id: String, val organizationId: String, val name: String)
data class AuthTokens(val accessToken: String, val refreshToken: String)

interface AuthClient {
    suspend fun exchange(code: String): Pair<AuthUser, AuthTokens>
    suspend fun refresh(refreshToken: String, organizationId: String? = null): AuthTokens
    suspend fun orgs(accessToken: String): List<AuthOrg>
}

class HttpAuthClient(
    private val config: AppConfig,
    private val http: HttpTransport,
) : AuthClient {
    private fun url(path: String) = "${config.edgeBaseUrl.trimEnd('/')}/$path"

    override suspend fun exchange(code: String): Pair<AuthUser, AuthTokens> {
        val body = JSONObject().put("code", code).toString().toByteArray()
        val resp = http.post(url("auth/exchange"), body)
        checkStatus(resp)
        val o = JSONObject(resp.text())
        val user = o.getJSONObject("user")
        val tokens = AuthTokens(o.getString("accessToken"), o.getString("refreshToken"))
        return AuthUser(user.getString("id"), user.optString("email").ifBlank { null }) to tokens
    }

    override suspend fun refresh(refreshToken: String, organizationId: String?): AuthTokens {
        val body = JSONObject().apply {
            put("refreshToken", refreshToken)
            if (organizationId != null) put("organizationId", organizationId)
        }.toString().toByteArray()
        val resp = http.post(url("auth/refresh"), body)
        checkStatus(resp)
        val o = JSONObject(resp.text())
        return AuthTokens(o.getString("accessToken"), o.getString("refreshToken"))
    }

    override suspend fun orgs(accessToken: String): List<AuthOrg> {
        val resp = http.get(url("auth/orgs"), mapOf("Authorization" to "Bearer $accessToken"))
        checkStatus(resp)
        val arr = JSONObject(resp.text()).getJSONArray("orgs")
        return (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            AuthOrg(o.getString("id"), o.getString("organizationId"), o.getString("name"))
        }
    }

    private fun checkStatus(resp: sh.zeron.android.sync.HttpResponse) {
        if (resp.code !in 200..299) {
            throw AppError.Auth(resp.code, resp.text().take(200))
        }
    }
}

class FakeAuthClient : AuthClient {
    var tokens = AuthTokens("fake-access", "fake-refresh")
    var orgs = listOf(AuthOrg("1", "org-1", "Org One"))
    override suspend fun exchange(code: String) = AuthUser("u1", "u@ex.com") to tokens
    override suspend fun refresh(refreshToken: String, organizationId: String?) = tokens
    override suspend fun orgs(accessToken: String) = orgs
}